### DISCUSSION TITLE
How do I debug ImageMagick on Linux?

### DISCUSSION THREAD

**Discussion by lbraby** [ImageMagick/ImageMagick#5282] [Help] [answered]
I have ImageMagick installed on my Linux OS but I'd like to know if there is any way to run it with a debugger like GDB. I'm having trouble understanding how the "-wavelet-denoise" command-line option works so I was hoping to be able to step through the C code to get a better understanding. Thanks for any help!

**Comment by urban-warrior** [comment, marked as answer]
ImageMagick is just an app.  Use `gdb` to step through the code.

**Reply by lbraby** [reply]
Ah ok, I was just wondering if there was a debug version of the build that I should use but I got it working with GDB so all is well. 

Do you by any chance have any additional information on the WaveletDenoiseImage function in MagickCore/visual-effects.c beyond what's in the code? I see that it is adapted from dcraw.c by David Coffin so no problem if there is only the code. However, if you know of a research paper or another resource that it is based on I would appreciate the link very much!

Thank you!

**Reply by urban-warrior** [reply]
Reviewing our notes, we only have one reference: [Denoising using wavelets](http://cs.haifa.ac.il/hagit/courses/seminars/wavelets/Presentations/Lecture09_Denoising.pdf).
